### PR TITLE
[prism] Fix downstream side input computations.

### PR DESCRIPTION
**Pull request by lostluck**
Fix error in prism's fusion handling. Re-instates fusion.

Instead of passing the consuming transform ID to get the downstream side inputs, we passed the PCollection ID, causing the computation to fail. Ultimately, the hangs were because there was a stage that depended on it's own produced side input.

Adds unit sufficient testing to avoid replicating the issue, and fixes the issue.
Exported fields on `link` and `fusionFacts` types to let them be compared for diffs by cmp. 
Adds additional debug logging for the resulting hang that was occurring.

(Long due to the unit tests, and necessary renames, the actual fix is a single line that I've commented to point out.)

---

Addendum: There was another bug in prism that when combined with fusion prevented one of the xlang tests from passing. In particular, the combiner lifting handler was not removing subtransforms recursively from lifted combines, leading to additional transforms being left in the graph.

This caused some interesting downstream mayhem: multiple producers for a single pcollection (pcollections only have one producer), and incorrect fusions, leading to the runner to hang, but *only* for certain topological orderings.

A unit test was added for the combine handler to validate the fix. 
Prism has also been made a bit more eager to fail when  it gets into those bad states, in particular when it notices that there is outstanding data to process, but nothing that would trigger it. Now prism crashes, which avoids indefinite hangs and timeouts.

----

Addendum 2.

Fixed a race condition in the pubsubx test, causing flakes.
Fixed an extra t.Run wrapping in the fileio.Match tests, which i assume caused flakes somehow.

Updated prisms bundle execution to use an errgroup to manage bundle failures, context cancellation, and limit parallelism, simplifying that code. This better addresses flakey handling of failures and pipeline completion.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/go/pkg/beam/runners/prism/internal/preprocess.go [hunk 1]
```diff
@@ -299,71 +299,71 @@ func checkForExpandCoderPattern(in, out string, comps *pipepb.Components) bool {
...
 	dssi := map[string]bool{}
 	for _, o := range comps.GetTransforms()[tID].GetOutputs() {
-		if facts.usedAsSideInput[o] {
+		if facts.UsedAsSideInput[o] {
 			dssi[o] = true
 		}
-		for _, consumer := range facts.pcolConsumers[o] {
-			cdssi := computeDownstreamSideInputs(consumer.global, comps, facts)
+		for _, consumer := range facts.PcolConsumers[o] {
+			cdssi := computeDownstreamSideInputs(consumer.Transform, comps, facts)
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/prism/internal/preprocess.go:362]: FYI, this line is where the bug occurred.

**Comment by lostluck** [comment]
I have confirmed that everything appears to be working, but something in my change has caused Prism to slow way down vs post adding the garbage collection. Investigating that now (I'm getting the same behavior locally).

**Comment by lostluck** [comment]
The root issue was that the subtransforms of "handled" composites weren't being pruned recursively. This lead to a weird topological sort dependent flakiness, causing the execution to eventually hit a dead end because a pcollection with data would have no consumers.

I've resolved that, and the hang goes away. I now need to clean up my debugging and see what's worth keeping.

File: sdks/go/pkg/beam/runners/prism/internal/preprocess.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/preprocess.go:398]: Well, the first bug.

**Comment by lostluck** [comment]
There were two other flaky tests that I've bundled in this as well, since they were conflating on runs.

Moving to the proper errgroup for parallelism limiting, and auto cancellation of other inprogress bundles on one bundle failure seems to have resolved the remaining flakiness (locally at least). The issue was the execution & error handling weren't uniformly handled, leading to a gap where an error could be missed between bundle completion (causing the element manager to exit successfully with all work completed), and actually handling an error (leading to false positive executions).

This got bigger than anticipated, so I'm happy to split it into component parts at reviewers discretion.

File: sdks/go/pkg/beam/io/fileio/read_test.go [hunk 2]
```diff
@@ -134,17 +134,15 @@ func TestReadMatches(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			t.Run(tt.name, func(t *testing.T) {
-				p, s := beam.NewPipelineWithRoot()
+			p, s := beam.NewPipelineWithRoot()
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/io/fileio/read_test.go:137]: good catch

**Review by riteshghorse — APPROVED** [review]
LGTM, thank you!

File: sdks/go/pkg/beam/runners/prism/internal/preprocess.go (hunk 1 above)
Comment (riteshghorse) [reply sdks/go/pkg/beam/runners/prism/internal/preprocess.go:393]: thanks for making the review short!

**Pull request merged**